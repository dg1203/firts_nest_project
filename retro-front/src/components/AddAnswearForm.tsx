import React, { useState, FormEvent, useContext } from 'react';
import BigInput from './BigInput';
import BigButton from './BigButton';
import BigSelect from './BigSelect';
import Checkbox from './Checkbox';
import Card from './Card';
import { QuestionDataInterface } from '../services/question/type';
import QuestionService from '../services/question/question.service';
import ToastContext from '../context/toast/toastContext';

interface AddAnswearPropsInterface {
  question: QuestionDataInterface;
  answearWasAdded: () => void;
}

const AddAnswearForm: React.FC<AddAnswearPropsInterface> = ({
  question,
  answearWasAdded
}) => {
  const questionService = new QuestionService();
  const toastContext = useContext(ToastContext);
  const [loading, setLoading] = useState<boolean>(false);
  const [answear, setAnswear] = useState<string>('');
  const [checkedData, setCheckedData] = useState<string[]>([]);
  const onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setLoading(true);
    try {
      const data = question.type === 'checkbox' ? checkedData : answear;
      const response = await questionService.addAnswear(data, question._id);
      toastContext.setMessage(response.data.message, 'success');
      setLoading(false);
      answearWasAdded();
    } catch (error) {
      toastContext.setMessage(error.response.data.message, 'error');
      setLoading(false);
    }
  };
  const onChangeCheckbox = (checked: boolean, index: number) => {
    if (checked) {
      const checkedCopy = checkedData;
      checkedCopy.push(question.options[index]);
      setCheckedData(checkedCopy);
    } else {
      const checkedCopy = checkedData.filter(
        (check: string) => check !== question.options[index]
      );
      setCheckedData(checkedCopy);
    }
  };
  return (
    <Card maxWidth={500}>
      <h2>{question.name}</h2>
      <form data-testid="form" onSubmit={event => onSubmit(event)}>
        {question.type === 'select' && (
          <BigSelect
            onChange={event => setAnswear(event.target.value)}
            required
            data-testid="select-input"
          >
            <option></option>
            {question.options.map((option: any, index: number) => (
              <option key={index} value={option}>
                {option}
              </option>
            ))}
          </BigSelect>
        )}
        {question.type === 'text' && (
          <BigInput
            onChange={event => setAnswear(event.target.value)}
            placeholder="Answear"
            value={answear}
            required
            data-testid="text-input"
          />
        )}
        {question.type === 'checkbox' &&
          question.options.map((option: string, index: number) => (
            <Checkbox
              label={option}
              key={index}
              idx={index}
              onChange={onChangeCheckbox}
            />
          ))}
        <BigButton isLoading={loading} isDisabled={loading}>
          Save
        </BigButton>
      </form>
    </Card>
  );
};

export default AddAnswearForm;
