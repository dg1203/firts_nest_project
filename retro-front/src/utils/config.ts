import axios from 'axios';

export const setAxiosConfig = (accessToken: string) => {
  axios.defaults.headers.common['Authorization'] = `Bearer ${accessToken}`;
  localStorage.setItem('accessToken', accessToken);
};
