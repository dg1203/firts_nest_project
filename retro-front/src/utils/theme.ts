interface ColorsInterface {
  blue: string;
  grey: string;
  white: string;
  red: string;
  black: string;
  green: string;
}

interface ThemeInterface {
  colors: ColorsInterface;
}

export const theme: ThemeInterface = {
  colors: {
    blue: '#2980b9',
    grey: '#bdc3c7',
    white: '#ffffff',
    red: '#c0392b',
    black: '#34495e',
    green: '#27ae60'
  }
};
