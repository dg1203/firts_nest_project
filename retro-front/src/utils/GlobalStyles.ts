import { createGlobalStyle } from "styled-components";
import { theme } from "./theme";

const GlobalStyles = createGlobalStyle`
  @import url(â€˜https://fonts.googleapis.com/css?family=Montserrat|Roboto');
  * {
    box-sizing: border-box;
  }
  body {
    margin: 0px;
    padding: 0px;
    font-family: Roboto, sans-serif;
    overflow-x: hidden;
  }
  a {
    color: ${theme.colors.blue};
    text-decoration: none;
  }
  form {
    width: 100%;
  }
`;

export default GlobalStyles;