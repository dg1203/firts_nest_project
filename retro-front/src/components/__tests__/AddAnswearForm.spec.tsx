import React from 'react';
import AddAnswearForm from '../AddAnswearForm';
import { ThemeProvider } from 'styled-components';
import { render, fireEvent } from '@testing-library/react';
import { theme } from '../../utils/theme';
import mockAxios from '../../__mocks__/axios';
import ToastContext from '../../context/toast/toastContext';
import { act } from 'react-dom/test-utils';
import exp from "constants";

const defaultProps = {
  question: {
    time: 3600,
    created: '20.05.2020',
    options: ['test', 'test1'],
    reply: [],
    _id: 'daldj219hfk',
    name: 'Test question?',
    code: 'testcode',
    open: true,
    type: 'text',
    __v: 0
  },
  answearWasAdded: jest.fn()
};

const defaultContext = {
  message: '',
  type: '',
  setMessage: (message: string, type: string) => ''
};

const renderComponent = (props: any) =>
  render(
    <ToastContext.Provider value={defaultContext}>
      <ThemeProvider theme={theme}>
        <AddAnswearForm {...props} />
      </ThemeProvider>
    </ToastContext.Provider>
  );

describe('<AddAnswearForm />', () => {
  it('To match snapshot', () => {
    const { asFragment } = renderComponent(defaultProps);
    expect(asFragment()).toMatchSnapshot();
  });
  describe('Test text input', () => {
    it('If type text should visible text input', () => {
      const { getByTestId } = renderComponent(defaultProps);
      expect(getByTestId('text-input')).toBeTruthy();
    });
    it('On change input value should set answear state', () => {
      const { getByTestId } = renderComponent(defaultProps);
      const input = getByTestId('text-input') as HTMLInputElement;
      expect(input.value).toBe('');
      fireEvent.change(input, {
        target: { value: 'answear' }
      });
      expect(input.value).toBe('answear');
    });
  });
  describe('Test select input', () => {
    it('If type select should visible select input', () => {
      const { getByTestId } = renderComponent({
        ...defaultProps,
        question: {
          ...defaultProps.question,
          type: 'select'
        }
      });
      expect(getByTestId('select-input')).toBeTruthy();
    });
    it('On change select value should set answear state', () => {
      const { getByTestId } = renderComponent({
        ...defaultProps,
        question: {
          ...defaultProps.question,
          type: 'select'
        }
      });
      const input = getByTestId('select-input') as HTMLSelectElement;
      expect(input.value).toBe('');
      fireEvent.change(input, {
        target: { value: 'test' }
      });
      expect(input.value).toBe('test');
    });
  });
  describe('Test checkbox input', () => {
    it('If type checkbox should visible checkbox inputs', () => {
      const { getAllByTestId } = renderComponent({
        ...defaultProps,
        question: {
          ...defaultProps.question,
          type: 'checkbox'
        }
      });
      expect(getAllByTestId('checkbox-input')).toBeTruthy();
      expect(getAllByTestId('checkbox-input').length).toBe(
        defaultProps.question.options.length
      );
    });
  });
  describe('Test form', () => {
    it('Check if form is exists', () => {
      const { getByTestId } = renderComponent(defaultProps);
      expect(getByTestId('form')).toBeTruthy();
    });
    it('Test submit form with success',  () => {
      const { getByTestId } = renderComponent(defaultProps);
      mockAxios.patch.mockImplementationOnce(() =>
        Promise.resolve({
          data: {
            message: 'Success'
          }
        })
      );
      fireEvent.submit(getByTestId('form'));
      expect(mockAxios.patch).toHaveBeenCalled();
    });
    it('Test submit form with error',  () => {
      const { getByTestId } = renderComponent(defaultProps);
      mockAxios.patch.mockImplementationOnce(() =>
          Promise.reject({
            response: {
              data: {
                message: 'Success'
              }
            }
          })
      );
      fireEvent.submit(getByTestId('form'));
      expect(mockAxios.patch).toHaveBeenCalled();
    });
  });
  describe('Test methods', () => {
    it('Test on change checkbox with checked true', () => {
      const { getAllByTestId } = renderComponent({
        ...defaultProps,
        question: {
          ...defaultProps.question,
          type: 'checkbox'
        }
      });
      const checkbox = getAllByTestId('checkbox-input')[0];
      fireEvent.change(checkbox, { checked: true, index: 0 });
    });
  });
});
