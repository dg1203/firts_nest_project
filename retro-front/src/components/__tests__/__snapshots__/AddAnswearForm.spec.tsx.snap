// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddAnswearForm /> To match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-AxgMl dqhpcW"
  >
    <h2>
      Test question?
    </h2>
    <form
      data-testid="form"
    >
      <input
        class="sc-AxjAm gqNWIl"
        data-testid="text-input"
        placeholder="Answear"
        required=""
        value=""
      />
      <button
        class="sc-AxirZ defsjn"
      >
        Save
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`<BigButton /> To match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-AxgMl dqhpcW"
  >
    <h2>
      Test question?
    </h2>
    <form
      data-testid="form"
    >
      <input
        class="sc-AxjAm gqNWIl"
        data-testid="text-input"
        placeholder="Answear"
        required=""
        value=""
      />
      <button
        class="sc-AxirZ defsjn"
      >
        Save
      </button>
    </form>
  </div>
</DocumentFragment>
`;
